### ISSUE TITLE
[$500] Workspace -Incorrect currency message flashes for second in the workspace where the  Bank account is added

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#36376]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.40-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause internal team
**Slack conversation:**

## Action Performed:
Precondition:
Fully verified bank account has been successfully added
Steps:
1. Navigate to Workspace Settings in which fully verified bank account is added
2. Click on Bank Account
3. click on Chats at the botton
4. Click on settings 

## Expected Result:
"The default currency can't be changed because this Workspace is linked to USD bank account" message should be present under Currency section

## Actual Result:
"All expenses on this workspace will be converted to this currency" flashes for a moment when Bank account is added

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/d560655f-c0dd-49ba-8455-3d89e6cefe05

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~015575d1ff4376d09e</li>
        <li>Upwork Job ID: 1757148556204711936</li>
        <li>Last Price Increase: 2024-02-12</li>
<li>Automatic offers: </li>
<ul>
<li>fedirjh | Reviewer | 0</li>
<li>bernhardoj | Contributor | 0</li>
</ul></li>
    </ul>
</details>

**Comment by Christinadobrzyn** [comment]
Hi @m-natarajan so sorry, I don't understand the reproduction steps in the OP.

Is the issue that the currency messaging on the workspace Bank Account page doesn't match the Profile page? 

https://github.com/Expensify/App/assets/51066321/61f00532-c41a-49fe-b796-615681235565

**Comment by bernhardoj** [comment]
Should have the same root cause as https://github.com/Expensify/App/issues/34919

**Comment by Christinadobrzyn** [comment]
oh nice -thanks @bernhardoj! Reached out in https://github.com/Expensify/App/issues/34919 to see if we can consolidate or if we should hold this until https://github.com/Expensify/App/issues/34919 is fixed.

**Comment by Christinadobrzyn** [comment]
Okay, we're going to put this on hold for https://github.com/Expensify/App/issues/34919#issuecomment-1942653681

**Comment by Christinadobrzyn** [comment]
hold for https://github.com/Expensify/App/issues/34919#issuecomment-1942653681

**Comment by Christinadobrzyn** [comment]
hold for https://github.com/Expensify/App/issues/34919#issuecomment-1942653681

**Comment by Christinadobrzyn** [comment]
Hold for - https://github.com/Expensify/App/issues/34919#issuecomment-1942653681

**Comment by Christinadobrzyn** [comment]
hold for - https://github.com/Expensify/App/issues/34919

**Comment by Christinadobrzyn** [comment]
hold for - https://github.com/Expensify/App/issues/34919

**Comment by Christinadobrzyn** [comment]
monitoring - https://github.com/Expensify/App/issues/34919

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The bank account page shows the connect option page instead of the all set page.

### What is the root cause of that problem?
It's a regression from https://github.com/Expensify/App/pull/29260 where we always clear the bank account data when we close the bank account page.

https://github.com/Expensify/App/blob/616d75f2e49be00f9bb0234cf2eb2cacf1522211/src/pages/ReimbursementAccount/ReimbursementAccountPage.js#L278-L287

So, when we reopen the bank account page, the page doesn't know we have a finished setup and shows the connect option page.

The purpose of that PR is to solve https://github.com/Expensify/App/issues/27901 (blinking issue). The issue is that when we open the BA page, it will first show the connect option page, a full-screen loader, and then the connect option page.

However, the PR itself is full of flaws.

First, we lost our ongoing BA setup whenever we closed the page.

Second, the second condition below always returns false.
https://github.com/Expensify/App/blob/616d75f2e49be00f9bb0234cf2eb2cacf1522211/src/pages/ReimbursementAccount/ReimbursementAccountPage.js#L247

```
_.has(this.props.reimbursementAccount, 'achData.currentStep');
```

Why? Undercore `has` can't accept a multi-level object property.
<img width="293" alt="image" src="https://github.com/Expensify/App/assets/50919443/ef581022-4d96-4908-a965-b29a9c923434">


### What changes do you think we should make in order to solve the problem?

Undo the changes from https://github.com/Expensify/App/pull/29260 and then we should have a new fix for the blinking issue https://github.com/Expensify/App/issues/27901.

To fix the blinking issue, we should understand the root cause first which is explained in the [blinking issue proposal](https://github.com/Expensify/App/issues/27901#issuecomment-1733395840). To put it short, we set the loading state of the BA onyx on mount, 
https://github.com/Expensify/App/blob/616d75f2e49be00f9bb0234cf2eb2cacf1522211/src/pages/ReimbursementAccount/ReimbursementAccountPage.js#L278-L280
https://github.com/Expensify/App/blob/616d75f2e49be00f9bb0234cf2eb2cacf1522211/src/pages/ReimbursementAccount/ReimbursementAccountPage.js#L266-L268

but it takes time, so we initially see the connect option page. After the onyx sends us the new data, a full-screen loading is shown, and after the loading is done, we see the connect option again.
https://github.com/Expensify/App/blob/616d75f2e49be00f9bb0234cf2eb2cacf1522211/src/pages/ReimbursementAccount/ReimbursementAccountPage.js#L402

From what I understand, we want to immediately show the loading indicator on the component mount. To achieve this, we should have a loading state/props that has an initial value of `true`. There are 2 ways I can think of to do that:

The first one is to have a new reimbursement account onyx connection with an undefined initial value and only takes the `isLoading` property. Here is what it looks like.

```js
isReimbursementAccountLoading: {
    key: ONYXKEYS.REIMBURSEMENT_ACCOUNT,
    initWithStoredValues: false,
    selector: (reimbursementAccount) => lodashGet(reimbursementAccount, 'isLoading'),
},
```

We set `initWithStoredValues: false,` so the onyx value is always undefined when the component is mounted. Because it's always undefined on mounted, we can set a default `isReimbursementAccountLoading` props value to `true` and then replaces all usages of `this.props.reimbursementAccount.isLoading` with `this.props.isReimbursementAccountLoading`.

However, `initWithStoredValues` is currently broken after the [onyx cache PR](https://github.com/Expensify/react-native-onyx/pull/280). So, to verify this solution, you can revert the PR locally, specifically this [line of code](https://github.com/Expensify/react-native-onyx/pull/280/files#diff-364fd7349a1af7b52091427c5a0734da8ff56057fbc7fe2be10734a06c97264cR40-R58), or see the onyx fix section below.

### What alternative solutions did you explore? (Optional)
The second one is to have a new local state called `isReimbursementAccountLoading` that has an initial value of `true`. All usages of `reimbursementAccount.isLoading` will be replaced with `isReimbursementAccountLoading`.

We will update `isLoading` state every time `reimbursementAccount.isLoading` is updated.
```js
const prevIsReimbursementAccountLoading = usePrevious(reimbursementAccount.isLoading);

useEffect(() => {
    if (reimbursementAccount.isLoading === prevIsReimbursementAccountLoading) return
    setIsLoading(reimbursementAccount.isLoading);
}, [prevIsReimbursementAccountLoading, reimbursementAccount.isLoading]);
```

Another alternative:
Both solutions above assume we want to keep the current behavior, that is by showing the loading indicator on component mount. This solution takes a different approach that is to show the loading indicator only if we never load the data yet.

Below is the solution
```js
if((!hasACHDataBeenLoaded && isLoading)
``` 

_Currently, we show the loading indicator if it's loading OR we don't have the data yet._
https://github.com/Expensify/App/blob/616d75f2e49be00f9bb0234cf2eb2cacf1522211/src/pages/ReimbursementAccount/ReimbursementAccountPage.js#L410

### Onyx fix
The root cause of the broken `initWithStoredValues` is that we [initialize all the onyx value from the cached value](https://github.com/Expensify/react-native-onyx/blob/14aed7bfba5668d1635bec5f5b99653c5a99916d/lib/withOnyx.js#L57-L97), even if we set `initWithStoredValues` as false.

To fix it, we should initialize the onyx key only if `initWithStoredValues` is true.
```diff
-if (
+if (mapping.initWithStoredValues &&
    ((value !== undefined
        && !Onyx.hasPendingMergeForKey(key))
    || mapping.allowStaleData)
) {
```

**Comment by bernhardoj** [comment]
@Christinadobrzyn https://github.com/Expensify/App/issues/34919 is closed based on this [comment](https://github.com/Expensify/App/issues/34919#issuecomment-2032619535), I think we should still fix the root cause of the issue to fix all of these issues: https://github.com/Expensify/App/issues/38230 https://github.com/Expensify/App/issues/39319

Since https://github.com/Expensify/App/issues/34919 has been really inactive, I decided to move my proposal here.

**Comment by Christinadobrzyn** [comment]
ah okay @fedirjh can you review https://github.com/Expensify/App/issues/36376#issuecomment-2032667831 proposal to see what you think?

It would make sense to fix the root of both https://github.com/Expensify/App/issues/38230 https://github.com/Expensify/App/issues/39319 at once if possible!

**Comment by Christinadobrzyn** [comment]
Hi @fedirjh can you take a peek at these proposals and @bernhardoj's [suggestion](https://github.com/Expensify/App/issues/36376#issuecomment-2032673029) to see if we can fix a few issues at once?

**Comment by bernhardoj** [comment]
> why can't we just use initialValue instead of using initWithStoredValues , I tested this approach, and it seems to work fine.

I haven't tried it yet, but if I can guess, it works because there is currently a bug in onyx where the `initialValue` is used even when we have the existing value of a boolean `false`.

**Comment by bernhardoj** [comment]
> there'll be this regression (which is also an issue with @bernhardoj 's proposal)

That's not a regression but an existing issue. Also, your proposal is almost identical to mine with the difference that you suggest to fix another issue.

**Comment by nkdengineer** [comment]
> That's not a regression but an existing issue.

@bernhardoj Can you confirm you can reproduce it even before applying your changes?

**Comment by nkdengineer** [comment]
@fedirjh I can't reproduce your video of the bug with my proposal, perhaps let me send you my branch to see if you can reproduce

> @nkdengineer What regression are you referring to?
Please try these steps (after applying @bernhardoj 's proposal)
1. Connect a new bank account (successfully)
2. Try clicking on `Bank account` in `WorkspaceInitialPage`, you'll the the loading indicator will never show (the root cause is explained in [my proposal](https://github.com/Expensify/App/issues/36376#issuecomment-2033678427))

The expectation (also existing behavior in production) is that the loading indicator should always show when first come to that page.

https://github.com/Expensify/App/assets/161821005/7bac11c9-654a-482e-b185-f5df43d40466

**Comment by bernhardoj** [comment]
It would be helpful if you could explain how my solution leads to the regression. We never clear the plaid data when unmounting on `main`. I notice about the plaid event issue in this [GH issue](https://github.com/Expensify/App/issues/33670) (the [test branch here](https://github.com/bernhardoj/Expensify/tree/fix/33670-unable-to-reaccess-plaid) is the proof), but I decided to not include it.

_BankAccounts.clearReimbursementAccount() never clears the plaid data_

**Comment by bernhardoj** [comment]
PR is ready

cc: @fedirjh

**Comment by nkdengineer** [comment]
@yuwenmemon There's discussion ongoing on [a regression](https://github.com/Expensify/App/issues/36376#issuecomment-2040293213) by the selected proposal, could we pause assignment here for a moment while we sort it out? Thanks!

**Comment by nkdengineer** [comment]
> It would be helpful if you could explain how my solution leads to the regression. 

@bernhardoj It happens after applying your change so I believe it is.

Do you agree it only starts happening after your change?

**Comment by bernhardoj** [comment]
I can't follow your repro steps as I can't seem to add a bank account from the new dot, so I add it from the old dot. You are saying that to solve the regression from my proposal, we should clear the plaid when unmounting.
```
BankAccounts.clearPlaid();
```

So, please explain which part of my proposal caused the regression that requires us to apply the code above instead of just saying that my proposal caused the regression.

**Comment by bernhardoj** [comment]
I think I finally understand the issue you mean. It's specifically this change.
<img width="1700" alt="image" src="https://github.com/Expensify/App/assets/50919443/af1bfdd7-0af6-41de-b1ba-a0d3c351b797">

When the plaid event is not cleared, `isLoading` variable is false. But we have another condition to show the loading, that is if `hasACHDataBeenLoaded` is `false`.
https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/pages/ReimbursementAccount/ReimbursementAccountPage.js#L436-L438

On `main`, `hasACHDataBeenLoaded` value is always false initially because of the bug as explained in my proposal, but then we fix it because the data is indeed loaded when we already have a connected bank account.

With or without my solution, the plaid event is never cleared. So, yeah, it's still an existing issue on the `main` where we don't clear the plaid event.

**Comment by bernhardoj** [comment]
@fedirjh Oh, I didn't know OpenReimbursementAccountPage clears the plaid data, but it's ONYXKEYS.PLAID_CURRENT_EVENT specifically that still exists.

**Comment by Christinadobrzyn** [comment]
Hi @bernhardoj @fedirjh - can you provide an update on this?

**Comment by bernhardoj** [comment]
@Christinadobrzyn hi, the PR is already [deployed to prod a week ago](https://github.com/Expensify/App/pull/39790#issuecomment-2073663171)

**Comment by Christinadobrzyn** [comment]
yikes! let's pay this out! I need to step away for an appt but I'll write up a payment summary so we can pay this out.

**Comment by Christinadobrzyn** [comment]
Payouts due:

- [X] Contributor: $500 @bernhardoj (paid in Upwork - https://www.upwork.com/nx/wm/offer/101753880)
- [X] Contributor+: $500 @fedirjh (paid in Upwork - https://www.upwork.com/nx/wm/offer/101753879)

Upwork job is [here](https://www.upwork.com/jobs/~015575d1ff4376d09e).

@bernhardoj or @fedirjh do we need a regression test?

**Comment by Christinadobrzyn** [comment]
nudge @bernhardoj or @fedirjh do we need a regression test?

**Comment by bernhardoj** [comment]
I think it's good to have a regression test. We can take the test step from the PR

Prerequisite: have a connected workspace bank account

1. Open the workspace that has the connected bank account
2. Open the workspace profile page
3. Verify the message below Default currency is "The default currency can't be changed..."
4. Open the bank account page
5. Verify you see a loading animation and there is no blinking
6. Close the bank account page
7. Open the workspace profile page again
8. Verify the message below Default currency is still the same as in step 3

**Comment by Christinadobrzyn** [comment]
Regression test - https://github.com/Expensify/Expensify/issues/394581

Closing this out, let me know if I'm missing anything!
